### ISSUE TITLE
Do not use homeserver to generate URL previews in desktop app

### ISSUE THREAD

**Issue by carlos22** [vector-im/riot-web#2953]
If the URL previews are enabled the riot-web ap 
`https://matrix.org/_matrix/media/r0/preview_url` (assuming that matrix.org is your home-server). This would not be necessary if the app runs inside of electron. The sesstion.webRequest API could be used instead. 
Benefits: For E2E rooms the user could have the URL preview without 'leaking' content of the chat to the home-server. The server doesn't need to proxy that much additional requests (can handle more users - theoretically)

See: https://github.com/electron/electron/pull/3640

Related to #2084

**Comment by carlos22** [comment]
Could we use the Platform abstraction to properly handle this?

**Comment by ara4n** [comment]
the reason that the server generates previews atm is to stop remote sites tracking IPs of clients.  This is particularly relevant for E2E users, I suspect.  We don't want to be in a situation where you have 1000 people sitting in an E2E room, and someone turns up and posts a url of https://evil.com or whatever, grabs a list of the IP addresses of all the connected users, and then attacks them.  There is currently no way for users to discover other users' IPs (and admins can only see the IPs connecting to their local HS).

**Comment by carlos22** [comment]
Ok that makes sense. For E2E direct-chats things are different, but matrix/riot does not distinguish between them. So it is all about how much you trust your homeserver...

**Comment by ara4n** [comment]
This very much feels like it should be a configurable option in the end. "Do you want to preview URLs via your server or via your client?  Do you want to preview URLs in E2E chats?"

**Comment by carlos22** [comment]
@ara4n agreed. As much as I want to keep settings simple, but in this case i think people should decide.

**Comment by tessgadwa** [comment]
From a UX perspective,  I also agree there should be a default but have it configurable.

**Comment by ara4n** [comment]
Alternatively, we could make it the sending client's problem (as per #3255).  It's an interesting question as to whether the preview should be telling the user what they will see if they click on the page (whilst leaking their IP, and hammering the server)... or telling the user what the sender intended the preview to be (as per #3255).

Unsure how to proceed here, really.

**Comment by carlos22** [comment]
Thats not a bad idea to offload the problem to the sender. Because most likely the sender opened the URL anyway? Some Apps in Andorid 6 do something similar, they take a screenshot and add it to the "share" processs.

**Comment by ara4n** [comment]
the only reason not to do it is trust, with spammers going and misrepresenting the target's contents in the thumbnail, as @CoffeeMatrix implies. this could be solved at a social layer (eg with reputation rules), but it's a bit of a shame to make for ourselves when the receiving client could just gen the thumbnail themselves (at the expense of revealing their IP and hammering the URL). luckily the code is pretty much the same whether it's receiver or sender genning the preview though.

**Comment by carlos22** [comment]
But whats the point? The spammers also cloud send URLs/Images with offensive content?
I think the protection of your IP and all the parameters that it sends to a server are more precocious than protection of images that are not correct. And if we are still talking about E2E Chats, than a spammer should not be verified and therefore no thumbnails should be showed?!

**Comment by lheckemann** [comment]
Generating the preview client-side also won't be possible in the browser because of XHR cross-origin restrictions, as far as I know.

**Comment by akontsevich** [comment]
Also vote to preview on the client!!!

**Comment by kevincox** [comment]
I would also like to see previews on the client. For example I am often sharing "secret links" in E2EE chats. I am not too worried about leaking my IP (at least in some rooms) but I would rather not share that URL with the server.

I think the best solution is to make the sender generate a preview. The receiver can then choose to display it, generate its own preview locally (not yet implemented) or ask the homeserver to generate the preview. I would send the preview as a reference to the original message so that it could be added by any user if the sender neglects to do so.

Random thoughts:

1. A user sending "spoofed" previews isn't really an issue, they can send me whatever rich content they want anyways. However users should be aware that the preview isn't "trusted".
2. Allowing IP leaks by default is a terrible idea. But there may be situations where it makes sense to allow a user to select receiver-client previews, generally on a per-room basis.
3. Some domains are trusted. For example maybe I don't mind `google.com` or `twitter.com` seeing my IP because I trust them to hide it from any attacker. 
4. Sender-sent previews are also nice because they provide a snapshot at the time of sending.
5. CORS makes it difficult to generate previews for web-based clients. Most websites don't allow this unfortunately. However a best-effort option could still be nice.
6. It would be possible to have different clients "attach" previews to messages. So maybe my desktop client can't generate a preview but someone else can and will attach it. This could be useful for web-based or minimal clients. This could even be done by a bot if desired (maybe a public room wants to do this so that URLs are reliably previewed, even if some users are linking non-CORS sites from a web client)

**Comment by teohhanhui** [comment]
Generating on the sender side is the right thing to do.

Keybase does the same: https://book.keybase.io/docs/chat/link-previews

**Comment by kevincox** [comment]
That link seems to show that keybase does it client side.

**Comment by teohhanhui** [comment]
> That link seems to show that keybase does it client side.

Yes, and the same as what you've suggested:

> I think the best solution is to make the sender generate a preview.

**Comment by HarHarLinks** [comment]
A common talking point and strong counter-argument to client-side previews appears to be
> People don't like leaking their IP address

Personally, I think this use case described by kevincox applies more often:
> For example I am often sharing "secret links" in E2EE chats. I am not too worried about leaking my IP (at least in some rooms) but I would rather not share that URL with the server.

This applies only limitedly (?? can you say this in english?) when you both are hosted on a trustworthy homeserver, e.g. all conversation parties are self-hosting. For the time being this is the rarer scenario, however you could argue that privacy concerned people would tend to do this anyway if they are able to.

In my opinion the strongest argument is that generating server-side takes any control out of your hands. Even if it is your own server, it will get associated with opening the link. When you are concerned about your IP leaking, it is reasonable to assume that you have taken steps for your whole PC, as you likely use more services than just element. For example you are likely using a VPN you trust. It seems reasonable that you are going to click a link you received eventually. In this case it would be better and safer to generate the preview from your own client.

I think https://github.com/vector-im/element-android/issues/3038 makes an interesting point: "[room] Administrators get to decide if the users of a room can individually enable/disable URL-Previews for their client. (defaulting to: disabled each)". In the context of this issue this could mean that room admins may disable server-side preview generation for all room participants.

**Comment by t3chguy** [comment]
Client side generation in the web can be problematic due to CORS preventing access to other domains unless they have `ACAO: *`

**Comment by Doomsdayrs** [comment]
Signal generates previews client side, and there seems to be no issues with that

**Comment by t3chguy** [comment]
@Doomsdayrs Signal doesn't look to be a webapp?

**Comment by Doomsdayrs** [comment]
> @Doomsdayrs Signal doesn't look to be a webapp?

It is not, but the android app shows that it is possible to get such a system done. Load the embed on user side and the user sends it out

**Comment by t3chguy** [comment]
Only Web apps are bound by ACAO headers, they're enforced by web browsers, not by native applications.
